### DISCUSSION TITLE
How to unit-test `asset_sensor`s

### DISCUSSION THREAD

**Discussion by OwenKephart** [dagster-io/dagster#12346] [Q&A] [answered]
Imagine you have an asset sensor defined as:

```
@asset_sensor(asset_key=AssetKey("foo"))
def my_sensor(context):
    ...
```

You may want to unit-test this sensor to make sure it behaves as expected. You can evaluate a tick of a sensor in the context of a unit test using `build_sensor_context`. However, asset sensors rely on querying for new events from the DagsterInstance, so you'll need to provide an `instance` parameter when building your context.

So the question becomes, how can you get a DagsterInstance that's suitable for unit tests?

**Comment by OwenKephart** [comment, marked as answer]
The solution starts off by using an ephemeral DagsterInstance (i.e. it will only exist for the duration of the test), and passing that into the `build_sensor_context` method:

```
from dagster import DagsterInstance, build_sensor_context

def test_my_sensor():
    with DagsterInstance.ephemeral() as instance:
        context = build_sensor_context(..., instance=instance)
        # note: we use list() here because my_sensor is actually a generator, so we need to exhaust the generator
        # to make sure our code runs
        result = list(my_sensor(context))
```

However, if you run this, your sensor code will not execute at all, as the sensor only runs the custom code when there are new AssetMaterialization events for the requested key. Because we just created this DagsterInstance, it will have no events in it.

To get past this, we'll need to materialize an asset. Depending on how complex your sensor code is, you can probably get away with just making sure that a materialization event exists for the key you care about, without worrying about how that event is created. This could look like:

```
from dagster import DagsterInstance, build_sensor_context, materialize_to_memory, asset

def test_my_sensor():
    @asset
    def foo():
        pass

    with DagsterInstance.ephemeral() as instance:
        # here, we materialize an asset with key `foo`, so that event will exist on our instance
        materialize_to_memory([foo], instance=instance)
        
        context = build_sensor_context(..., instance=instance)
        result = list(my_sensor(context))
```

For complex use cases, you may want to import the actual definition of the asset you care about and pass that into `materialize_to_memory`, rather than mocking it in the test.

**Comment by isaac-jordan** [comment]
Note: You can make your tests significantly faster if you use Dagster's built-in `instance_for_test`:

```
from dagster import instance_for_test
...
def test_my_sensor():
    ...
    overrides = {
        "run_launcher": {
            "module": "dagster._core.test_utils",
            "class": "MockedRunLauncher",
        },
    }
    with instance_for_test(overrides=overrides) as instance:
        ...
```
